src/commonMain/kotlin/fr/acinq/eclair/payment/IncomingPaymentHandler.kt

```diff
@@ -93,6 +93,54 @@ class IncomingPaymentHandler(val nodeParams: NodeParams, val walletParams: Walle
         return pr
     }
 
+    /**
+     * Save the "received-with" details of an incoming amount.
+     *
+     * - for a pay-to-open origin, we only save the id of the channel that was created for this payment.
+     * - for a swap-in origin, a new incoming payment must be created. We use the channel id to generate the payment's preimage.
+     * - for unknown origin, the amount is handled as a swap-in coming from an unknown address.
+     */
+    suspend fun process(channelId: ByteVector32, action: ChannelAction.Storage.StoreIncomingAmount) {
+        when (action.origin) {
+            null -> {
+                logger.warning { "incoming amount with empty origin, we store only minimal information" }
+                val fakePreimage = channelId.sha256()
+                db.addAndReceivePayment(
+                    preimage = fakePreimage,
+                    origin = IncomingPayment.Origin.SwapIn(address = ""),
+                    amount = action.amount,
+                    receivedWith = IncomingPayment.ReceivedWith.NewChannel(fees = 0.msat, channelId = channelId)
+                )
+            }
+            is ChannelOrigin.PayToOpenOrigin -> {
+                if (db.getIncomingPayment(action.origin.paymentHash) != null) {
+                    db.receivePayment(
+                        paymentHash = action.origin.paymentHash,
+                        amount = 0.msat, // do not update the amount, it's already set by the main payment handler.
+                        receivedWith = IncomingPayment.ReceivedWith.NewChannel(
+                            fees = action.origin.fee.toMilliSatoshi(),
+                            channelId = channelId
+                        )
+                    )
+                } else {
+                    logger.warning { "c:$channelId ignored pay-to-open origin action, there are no payments in db for payment_hash=${action.origin.paymentHash}" }
```

Comment:
It seems that the node/channel context is handled somewhere else, so maybe we can get rid of it here?
```suggestion
                    logger.warning { "ignored pay-to-open origin action, there are no payments in db for payment_hash=${action.origin.paymentHash}" }
```